crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -0,0 +1,1714 @@
+        instance_limits.memory_reservation_size = u64::try_from(round_up_to_pow2(
+            usize::try_from(instance_limits.memory_reservation_size).unwrap(),
+            WASM_PAGE_SIZE as usize,
+        ))
+        .unwrap();
+
+        // Cap the memory reservation size to 8 GiB (maximum 4 GiB accessible + 4 GiB of guard region)
+        instance_limits.memory_reservation_size =
+            min(instance_limits.memory_reservation_size, 0x200000000);
+
+        // The maximum module memory page count cannot exceed 65536 pages
+        if module_limits.memory_pages > 0x10000 {
+            return Err(format!(
+                "module memory page limit of {} exceeds the maximum of 65536",
+                module_limits.memory_pages
+            ));
+        }
+
+        // The maximum module memory page count cannot exceed the memory reservation size
+        if (module_limits.memory_pages * WASM_PAGE_SIZE) as u64
+            > instance_limits.memory_reservation_size
+        {
+            return Err(format!(
+                "module memory page limit of {} pages exeeds the memory reservation size limit of {} bytes",
+                module_limits.memory_pages,
+                instance_limits.memory_reservation_size
+            ));
+        }
+
+        let instances = InstancePool::new(&module_limits, &instance_limits)?;
+        let stacks = StackPool::new(&instance_limits, stack_size)?;
+
+        #[cfg(all(feature = "uffd", target_os = "linux"))]
+        let _fault_handler = PageFaultHandler::new(&instances, &stacks)?;
+
+        Ok(Self {
+            strategy,
+            module_limits,
+            instance_limits,
+            instances: mem::ManuallyDrop::new(instances),
+            stacks: mem::ManuallyDrop::new(stacks),
+            #[cfg(all(feature = "uffd", target_os = "linux"))]
+            _fault_handler,
+        })
+    }
+}
+
+impl Drop for PoolingInstanceAllocator {
+    fn drop(&mut self) {
+        // Manually drop the pools before the fault handler (if uffd is enabled)
+        // This ensures that any fault handler thread monitoring the pool memory terminates
+        unsafe {
+            mem::ManuallyDrop::drop(&mut self.instances);
+            mem::ManuallyDrop::drop(&mut self.stacks);
+        }
+    }
+}
+
+unsafe impl InstanceAllocator for PoolingInstanceAllocator {
+    fn validate_module(&self, translation: &ModuleTranslation) -> Result<(), String> {
+        self.module_limits.validate_module(&translation.module)
+    }
+
+    fn adjust_tunables(&self, tunables: &mut Tunables) {
+        let memory_reservation_size = self.instance_limits.memory_reservation_size;
+
+        // For reservation sizes larger than 4 GiB, use a guard region to elide bounds checks
+        if memory_reservation_size >= 0x100000000 {
+            tunables.static_memory_bound = 0x10000; // in Wasm pages
+            tunables.static_memory_offset_guard_size = memory_reservation_size - 0x100000000;
```

Comment:
I can do that.

I've also been toying with removing the `memory_reservation_size` knob in the pooling instance allocator limits and instead relying on `static_memory_bound` when creating the pooling instance allocator.  A static memory bound of zero would be an error when using the pooling instance allocator and it would still be treated as the upper bound for all memories (and thus are static).

This would mean the same sort of checks done in `Config::async_stack_size` to see if the strategy was already configured when setting `static_memory_bound` would need to be performed, but that's relatively minor.

Reply:
Except now I just remembered why I quashed that idea: the 32-bit default for `static_memory_bound` is way too large to accommodate the pooling instance allocator (1 GiB vs 10 MiB).

Reply:
Yeah I don't think we need to be super elegant with these knobs, they're quite advanced and will likely rarely get tweaked by folks other than us anyway. I think it's good to have notes about the effects but no need to make them the most ergonomic.